feat (183): User Activity Map

techeclecticdesign opened the pull request:
Colors defined in tailwind.config in existing daisyui themes.
Responsiveness by listening to window resize event and adjusting size to 1 of 3 breakpoints.
DaisyUI tooltips.
Large amount of code for a front end component, but nothing too resource hungry that I can tell.
Open to any adjustments to make this best fit with the UnlockEdv2 platform.

![image](https://github.com/UnlockedLabs/UnlockEdv2/assets/130294759/021210b4-f9e2-4696-9d34-ac51e6a5f0ae)
![image](https://github.com/UnlockedLabs/UnlockEdv2/assets/130294759/454b5fb7-0e87-4825-8797-4545546fbe9f)

PThorpe92 commented:
That looks awesome! Will take a look at this tomorrow. Good stuff man

calisio commented:
This is SO great!! I'm looking at it today too :) Thank you for tackling this!

calisio commented on resources/js/Components/UserActivityMap.tsx:
```diff
@@ -0,0 +1,371 @@
+        return years;
+    };
+
+    const fetchActivityData = async () => {
+        const startDate = new Date(subtractYear(yearEnd));
+        const response = await axios.get(
+            "/api/v1/user-activity-map/" +
+                user.id +
+                "?start_date=" +
+                startDate.toISOString().split("T")[0] +
+                "&end_date=" +
+                yearEnd.toISOString().split("T")[0],
+        );
+        const data = await response.data;
+        setActivityData(data);
+    };
```
Rather than use axios.get, we have been using a react hook to fetch data called useSWR (https://swr.vercel.app/). You can see an example in `Users.tsx` (and other files like provider platforms, user activity, etc). With this you can also handle loading & errors and render something accordingly.
This is totally my bad because I completely forgot to add this into the `CONTRIBUTING.md` file. I'll do that soon :)

calisio commented on resources/js/Components/UserActivityMap.tsx:
```diff
@@ -0,0 +1,371 @@
+        setAggregateActivityTime(Math.floor(aggregateActivityTime / 3600));
+    }, [aggregateActivityTime]);
+
+    /* add empty cells for days of week after activity range */
+    for (i = dateCount.getUTCDay(); i < 7; i++) {
+        tableData.push(
+            <td className="block" key={i + 7}>
+                <div className={`${nodeSizes[size]}`}></div>
+            </td>,
+        );
+    }
```
do you think for this we could just add grey cells until the end of the year? I just was thinking this because its similar to how github does it. Open to what you think looks best though. @PThorpe92 do you have any thoughts / feelings towards this?
<img width="951" alt="Screenshot 2024-04-17 at 11 20 00 AM" src="https://github.com/UnlockedLabs/UnlockEdv2/assets/89868388/076f3d85-554a-46da-b26b-6e7df979345a">
<img width="832" alt="Screenshot 2024-04-17 at 11 21 30 AM" src="https://github.com/UnlockedLabs/UnlockEdv2/assets/89868388/833051bf-4bcb-4adb-bacc-2df4a4d683db">

PThorpe92 replied:
I like that :+1:  I think grey by default, so there is a year's worth at all times is the way to go.

techeclecticdesign commented:
Thank you for the feedback, if there's anything else let me know. I will have your recommended changes complete by this weekend.

calisio commented:
That sounds great. Would you also be able to put one more node in the legend that is just a grey square? For a total of 5 squares

techeclecticdesign commented:
I believe this is done now.  In addition to the recommended fixes, I added months at top, added a  semi-transparent border to the activity cells (similar to how github looks), and in general improved behavior and appearance.  Let me know if there are any other adjustments needed.

calisio reviewed (approved):
This looks great! Thanks for making those changes 🙂
